// @twurple/api/es/api/helix/moderation/HelixBanUser.mjs
import { __decorate } from "tslib";
import { Enumerable, mapNullable } from '@d-fischer/shared-utils';
import { DataObject, rawDataSymbol, rtfm } from '@twurple/common';
/**
 * Information about a user who has been banned/timed out.
 *
 * @hideProtected
 */
let HelixBanUser = class HelixBanUser extends DataObject {
    /** @private */
    constructor(data, broadcasterId, expiryTimestamp, client) {
        super(data);
        this._broadcasterId = broadcasterId;
        this._expiryTimestamp = expiryTimestamp;
        this._client = client;
    }
    /**
     * The ID of the broadcaster whose chat room the user was banned/timed out from chatting in.
     *
     * @deprecated As this is the result of a ban action or list request that takes the broadcaster ID as a parameter,
     * use that directly instead.
     */
    get broadcasterId() {
        return this._broadcasterId;
    }
    /**
     * The date and time that the ban/timeout was created.
     */
    get creationDate() {
        return new Date(this[rawDataSymbol].created_at);
    }
    /**
     * The date and time that the timeout will end. Is `null` if the user was banned instead of put in a timeout.
     */
    get expiryDate() {
        return mapNullable(this._expiryTimestamp, ts => new Date(ts));
    }
    /**
     * @deprecated Use {@link HelixBanUser#expiryDate} instead.
     */
    get endDate() {
        return this.expiryDate;
    }
    /**
     * The ID of the moderator that banned or put the user in the timeout.
     */
    get moderatorId() {
        return this[rawDataSymbol].moderator_id;
    }
    /**
     * Retrieves more information about the moderator that banned or put the user in the timeout.
     */
    async getModerator() {
        return (await this._client.users.getUserById(this[rawDataSymbol].moderator_id));
    }
    /**
     * The ID of the user that was banned or put in a timeout.
     */
    get userId() {
        return this[rawDataSymbol].user_id;
    }
    /**
     * Fetches more info about the user that was banned or put in a timeout.
     */
    async getUser() {
        return (await this._client.users.getUserById(this[rawDataSymbol].user_id));
    }
};
__decorate([
    Enumerable(false)
], HelixBanUser.prototype, "_client", void 0);
__decorate([
    Enumerable(false)
], HelixBanUser.prototype, "_broadcasterId", void 0);
__decorate([
    Enumerable(false)
], HelixBanUser.prototype, "_expiryTimestamp", void 0);
HelixBanUser = __decorate([
    rtfm('api', 'HelixBanUser', 'userId')
], HelixBanUser);
export { HelixBanUser };
