// @twurple/api/es/interfaces/helix/moderation.external.mjs
import { extractUserId } from '@twurple/common';
/** @private */
export function createModerationUserListQuery(channel, filter) {
    return {
        broadcaster_id: extractUserId(channel),
        user_id: filter === null || filter === void 0 ? void 0 : filter.userId
    };
}
/** @private */
export function createModeratorModifyQuery(broadcaster, user) {
    return {
        broadcaster_id: extractUserId(broadcaster),
        user_id: extractUserId(user)
    };
}
/** @private */
export function createAutoModProcessBody(user, msgId, allow) {
    return {
        user_id: extractUserId(user),
        msg_id: msgId,
        action: allow ? 'ALLOW' : 'DENY'
    };
}
/** @private */
export function createAutoModSettingsBody(data) {
    return {
        overall_level: data.overallLevel,
        aggression: data.aggression,
        bullying: data.bullying,
        disability: data.disability,
        misogyny: data.misogyny,
        race_ethnicity_or_religion: data.raceEthnicityOrReligion,
        sex_based_terms: data.sexBasedTerms,
        sexuality_sex_or_gender: data.sexualitySexOrGender,
        swearing: data.swearing
    };
}
/** @private */
export function createBanUserBody(data) {
    return {
        data: {
            duration: data.duration,
            reason: data.reason,
            user_id: extractUserId(data.userId)
        }
    };
}
/** @private */
export function createUpdateShieldModeStatusBody(activate) {
    return {
        is_active: activate
    };
}
