// @twurple/api/lib/api/helix/channel/HelixChannelApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixChannelApi = void 0;
const tslib_1 = require("tslib");
const api_call_1 = require("@twurple/api-call");
const common_1 = require("@twurple/common");
const channel_external_1 = require("../../../interfaces/helix/channel.external");
const generic_external_1 = require("../../../interfaces/helix/generic.external");
const BaseApi_1 = require("../../BaseApi");
const HelixPaginatedRequest_1 = require("../HelixPaginatedRequest");
const HelixPaginatedResult_1 = require("../HelixPaginatedResult");
const HelixPagination_1 = require("../HelixPagination");
const HelixUserRelation_1 = require("../relations/HelixUserRelation");
const HelixChannel_1 = require("./HelixChannel");
const HelixChannelEditor_1 = require("./HelixChannelEditor");
/**
 * The Helix API methods that deal with channels.
 *
 * Can be accessed using `client.channels` on an {@link ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const channel = await api.channels.getChannelInfoById('125328655');
 * ```
 *
 * @meta category helix
 * @meta categorizedTitle Channels
 */
let HelixChannelApi = class HelixChannelApi extends BaseApi_1.BaseApi {
    /**
     * Retrieves the channel data for the given user.
     *
     * @param user The user you want to get channel info for.
     */
    async getChannelInfoById(user) {
        var _a;
        const channels = await this.getChannelInfoByIds([user]);
        return (_a = channels[0]) !== null && _a !== void 0 ? _a : null;
    }
    /**
     * Retrieves the channel data for the given users.
     *
     * @param users The users you want to get channel info for.
     */
    async getChannelInfoByIds(users) {
        const userIds = users.map(common_1.extractUserId);
        const result = await this._client.callApi({
            type: 'helix',
            url: 'channels',
            query: (0, generic_external_1.createSingleKeyQuery)('broadcaster_id', userIds)
        });
        return result.data.map(data => new HelixChannel_1.HelixChannel(data, this._client));
    }
    /**
     * Retrieves the channel data for the given user.
     *
     * @param user The user you want to get channel info for.
     *
     * @deprecated Use `getChannelInfoById` instead.
     */
    async getChannelInfo(user) {
        return await this.getChannelInfoById(user);
    }
    /**
     * Updates the given user's channel data.
     *
     * @param user The user you want to update channel info for.
     * @param data The channel info to set.
     */
    async updateChannelInfo(user, data) {
        await this._client.callApi({
            type: 'helix',
            url: 'channels',
            method: 'PATCH',
            scope: 'channel:manage:broadcast',
            query: (0, api_call_1.createBroadcasterQuery)(user),
            jsonBody: (0, channel_external_1.createChannelUpdateBody)(data)
        });
    }
    /**
     * Starts a commercial on a channel.
     *
     * @param broadcaster The broadcaster on whose channel the commercial is started.
     * @param length The length of the commercial, in seconds.
     */
    async startChannelCommercial(broadcaster, length) {
        await this._client.callApi({
            type: 'helix',
            url: 'channels/commercial',
            method: 'POST',
            scope: 'channel:edit:commercial',
            jsonBody: (0, channel_external_1.createChannelCommercialBody)(broadcaster, length)
        });
    }
    /**
     * Retrieves a list of users who have editor permissions on your channel.
     *
     * @param broadcaster The broadcaster to retreive the editors for.
     */
    async getChannelEditors(broadcaster) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'channels/editors',
            scope: 'channel:read:editors',
            query: (0, api_call_1.createBroadcasterQuery)(broadcaster)
        });
        return result.data.map(data => new HelixChannelEditor_1.HelixChannelEditor(data, this._client));
    }
    /**
     * Retrieves a list of VIPs in a channel.
     *
     * @param broadcaster The owner of the channel to get VIPs for.
     * @param pagination
     *
     * @expandParams
     */
    async getVips(broadcaster, pagination) {
        const response = await this._client.callApi({
            type: 'helix',
            url: 'channels/vips',
            scope: 'channel:read:vips',
            query: {
                ...(0, api_call_1.createBroadcasterQuery)(broadcaster),
                ...(0, HelixPagination_1.createPaginationQuery)(pagination)
            }
        });
        return (0, HelixPaginatedResult_1.createPaginatedResult)(response, HelixUserRelation_1.HelixUserRelation, this._client);
    }
    /**
     * Creates a paginator for VIPs in a channel.
     *
     * @param broadcaster The owner of the channel to get VIPs for.
     */
    getVipsPaginated(broadcaster) {
        return new HelixPaginatedRequest_1.HelixPaginatedRequest({
            url: 'channels/vips',
            scope: 'channel:read:vips',
            query: (0, api_call_1.createBroadcasterQuery)(broadcaster)
        }, this._client, data => new HelixUserRelation_1.HelixUserRelation(data, this._client));
    }
    /**
     * Checks the VIP status of a list of users in a channel.
     *
     * @param broadcaster The owner of the channel to check VIP status in.
     * @param users The users to check.
     */
    async checkVipForUsers(broadcaster, users) {
        const response = await this._client.callApi({
            type: 'helix',
            url: 'channels/vips',
            scope: 'channel:read:vips',
            query: (0, generic_external_1.createChannelUsersCheckQuery)(broadcaster, users)
        });
        return response.data.map(data => new HelixUserRelation_1.HelixUserRelation(data, this._client));
    }
    /**
     * Checks the VIP status of a user in a channel.
     *
     * @param broadcaster The owner of the channel to check VIP status in.
     * @param user The user to check.
     */
    async checkVipForUser(broadcaster, user) {
        const userId = (0, common_1.extractUserId)(user);
        const result = await this.checkVipForUsers(broadcaster, [userId]);
        return result.some(rel => rel.id === userId);
    }
    /**
     * Adds a VIP to the broadcaster’s chat room.
     *
     * @param broadcaster The broadcaster that’s granting VIP status to the user. This ID must match the user ID in the access token.
     * @param user The user to add as a VIP in the broadcaster’s chat room.
     */
    async addVip(broadcaster, user) {
        await this._client.callApi({
            type: 'helix',
            url: 'channels/vips',
            method: 'POST',
            scope: 'channel:manage:vips',
            query: (0, channel_external_1.createChannelVipUpdateQuery)(broadcaster, user)
        });
    }
    /**
     * Removes a VIP from the broadcaster’s chat room.
     *
     * @param broadcaster The broadcaster that’s removing VIP status from the user. This ID must match the user ID in the access token.
     * @param user The user to remove as a VIP from the broadcaster’s chat room.
     */
    async removeVip(broadcaster, user) {
        await this._client.callApi({
            type: 'helix',
            url: 'channels/vips',
            method: 'DELETE',
            scope: 'channel:manage:vips',
            query: (0, channel_external_1.createChannelVipUpdateQuery)(broadcaster, user)
        });
    }
};
HelixChannelApi = tslib_1.__decorate([
    (0, common_1.rtfm)('api', 'HelixChannelApi')
], HelixChannelApi);
exports.HelixChannelApi = HelixChannelApi;
