// ircv3/lib/Message/MessageTypes/Numerics/Reply352WhoReply.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reply352WhoReply = void 0;
const tslib_1 = require("tslib");
const Message_1 = require("../../Message");
const MessageDefinition_1 = require("../../MessageDefinition");
const WhoQuery_1 = require("../Commands/WhoQuery");
let Reply352WhoReply = class Reply352WhoReply extends Message_1.Message {
    /**
     * Checks whether the found user is /away.
     */
    get isAway() {
        return this.params.flags.includes('G');
    }
    /**
     * Checks whether the found user is an IRCOp.
     */
    get isOper() {
        return this.params.flags.includes('*');
    }
    /**
     * Checks whether the found user is a bot.
     */
    get isBot() {
        return this.params.flags.includes('B');
    }
    isResponseTo(originalMessage) {
        return originalMessage instanceof WhoQuery_1.WhoQuery;
    }
};
tslib_1.__decorate([
    (0, MessageDefinition_1.MessageParamDefinition)()
], Reply352WhoReply.prototype, "me", void 0);
tslib_1.__decorate([
    (0, MessageDefinition_1.MessageParamDefinition)()
], Reply352WhoReply.prototype, "channel", void 0);
tslib_1.__decorate([
    (0, MessageDefinition_1.MessageParamDefinition)()
], Reply352WhoReply.prototype, "user", void 0);
tslib_1.__decorate([
    (0, MessageDefinition_1.MessageParamDefinition)()
], Reply352WhoReply.prototype, "host", void 0);
tslib_1.__decorate([
    (0, MessageDefinition_1.MessageParamDefinition)()
], Reply352WhoReply.prototype, "server", void 0);
tslib_1.__decorate([
    (0, MessageDefinition_1.MessageParamDefinition)()
], Reply352WhoReply.prototype, "nick", void 0);
tslib_1.__decorate([
    (0, MessageDefinition_1.MessageParamDefinition)()
], Reply352WhoReply.prototype, "flags", void 0);
tslib_1.__decorate([
    (0, MessageDefinition_1.MessageParamDefinition)({
        trailing: true
    })
], Reply352WhoReply.prototype, "hopsAndRealName", void 0);
Reply352WhoReply = tslib_1.__decorate([
    (0, MessageDefinition_1.MessageType)('352')
], Reply352WhoReply);
exports.Reply352WhoReply = Reply352WhoReply;
