// ircv3/es/Message/MessageTypes/Numerics/Reply352WhoReply.mjs
import { __decorate } from "tslib";
import { Message } from "../../Message.mjs";
import { MessageParamDefinition, MessageType } from "../../MessageDefinition.mjs";
import { WhoQuery } from "../Commands/WhoQuery.mjs";
let Reply352WhoReply = class Reply352WhoReply extends Message {
    /**
     * Checks whether the found user is /away.
     */
    get isAway() {
        return this.params.flags.includes('G');
    }
    /**
     * Checks whether the found user is an IRCOp.
     */
    get isOper() {
        return this.params.flags.includes('*');
    }
    /**
     * Checks whether the found user is a bot.
     */
    get isBot() {
        return this.params.flags.includes('B');
    }
    isResponseTo(originalMessage) {
        return originalMessage instanceof WhoQuery;
    }
};
__decorate([
    MessageParamDefinition()
], Reply352WhoReply.prototype, "me", void 0);
__decorate([
    MessageParamDefinition()
], Reply352WhoReply.prototype, "channel", void 0);
__decorate([
    MessageParamDefinition()
], Reply352WhoReply.prototype, "user", void 0);
__decorate([
    MessageParamDefinition()
], Reply352WhoReply.prototype, "host", void 0);
__decorate([
    MessageParamDefinition()
], Reply352WhoReply.prototype, "server", void 0);
__decorate([
    MessageParamDefinition()
], Reply352WhoReply.prototype, "nick", void 0);
__decorate([
    MessageParamDefinition()
], Reply352WhoReply.prototype, "flags", void 0);
__decorate([
    MessageParamDefinition({
        trailing: true
    })
], Reply352WhoReply.prototype, "hopsAndRealName", void 0);
Reply352WhoReply = __decorate([
    MessageType('352')
], Reply352WhoReply);
export { Reply352WhoReply };
