// @twurple/api/lib/interfaces/helix/generic.external.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createChannelUsersCheckQuery = exports.createGetByIdsQuery = exports.createModeratorActionQuery = exports.createSingleKeyQuery = void 0;
const common_1 = require("@twurple/common");
/** @private */
function createSingleKeyQuery(key, value) {
    return { [key]: value };
}
exports.createSingleKeyQuery = createSingleKeyQuery;
/** @private */
function createModeratorActionQuery(broadcaster, moderator) {
    return {
        broadcaster_id: (0, common_1.extractUserId)(broadcaster),
        moderator_id: (0, common_1.extractUserId)(moderator)
    };
}
exports.createModeratorActionQuery = createModeratorActionQuery;
/** @private */
function createGetByIdsQuery(broadcaster, rewardIds) {
    return {
        broadcaster_id: (0, common_1.extractUserId)(broadcaster),
        id: rewardIds
    };
}
exports.createGetByIdsQuery = createGetByIdsQuery;
/** @private */
function createChannelUsersCheckQuery(broadcaster, users) {
    return {
        broadcaster_id: (0, common_1.extractUserId)(broadcaster),
        user_id: users.map(common_1.extractUserId)
    };
}
exports.createChannelUsersCheckQuery = createChannelUsersCheckQuery;
