// @twurple/api/lib/interfaces/helix/poll.external.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPollEndBody = exports.createPollBody = void 0;
const common_1 = require("@twurple/common");
/** @private */
function createPollBody(broadcaster, data) {
    var _a, _b;
    return {
        broadcaster_id: (0, common_1.extractUserId)(broadcaster),
        title: data.title,
        choices: data.choices.map(title => ({ title })),
        duration: data.duration,
        bits_voting_enabled: data.bitsPerVote != null,
        bits_per_vote: (_a = data.bitsPerVote) !== null && _a !== void 0 ? _a : 0,
        channel_points_voting_enabled: data.channelPointsPerVote != null,
        channel_points_per_vote: (_b = data.channelPointsPerVote) !== null && _b !== void 0 ? _b : 0
    };
}
exports.createPollBody = createPollBody;
/** @private */
function createPollEndBody(broadcaster, id, showResult) {
    return {
        broadcaster_id: (0, common_1.extractUserId)(broadcaster),
        id,
        status: showResult ? 'TERMINATED' : 'ARCHIVED'
    };
}
exports.createPollEndBody = createPollEndBody;
